OpenLiberty 20.0.0.12-beta post to staging (#832)
OpenLiberty/blogs: posts/2020-10-23-beta-20.0.0.12.adoc

MicroProfile 4.0 brings the following feature updates to Open Liberty:

* Config 2.0
* Fault Tolerance 3.0
* Metrics 3.0
* Health 3.0
* Open API 2.0
* JWT 1.2
* Open Tracing 2.0

To enable the new beta features in your app, add them to your `pom.xml`:

[source, xml]
----
    <dependencies>
        <dependency>
            <groupId>org.eclipse.microprofile</groupId>
            <artifactId>microprofile</artifactId>
            <version>4.0</version>
            <type>pom</type>
            <scope>provided</scope>
        </dependency>
    </dependencies>
----

For more information about the MicroProfile updates included in this milestone, visit the link:https://download.eclipse.org/microprofile/microprofile-4.0-M1/microprofile-spec-4.0-M1.html#microprofile4.0-M1[release page].

[#conf]
=== Support for MicroProfile Config 2.0

> link:https://github.com/eclipse/microprofile-config[MicroProfile] Config is a specification which OpenLiberty implements to allow application configuration from different locations (ConfigSources). In the previous versions, OpenLiberty has implemented the specification in its own codebase. However, for MicroProfile Config 2.0 it consumes link:https://github.com/smallrye/smallrye-config[SmallRye's implementation] of the specification.

Reviewer: ```suggestion
link:https://github.com/eclipse/microprofile-config[MicroProfile Config] is a specification which Open Liberty implements to allow application configuration from different locations (ConfigSources). In the previous versions, Open Liberty has implemented the specification in its own codebase. However, for MicroProfile Config 2.0 it consumes link:https://github.com/smallrye/smallrye-config[SmallRye's implementation] of the specification.
```

suggest changing "OpenLiberty" to "Open Liberty" and extend the link for "MicroProfile Config"

Author: This has been changed, thanks.

> link:https://github.com/eclipse/microprofile-config[MicroProfile] Config is a specification which OpenLiberty implements to allow application configuration from different locations (ConfigSources). In the previous versions, OpenLiberty has implemented the specification in its own codebase. However, for MicroProfile Config 2.0 it consumes link:https://github.com/smallrye/smallrye-config[SmallRye's implementation] of the specification.

Reviewer 2: ```suggestion
link:https://github.com/eclipse/microprofile-config[MicroProfile Config] is a specification which OpenLiberty implements to allow application configuration from different locations (ConfigSources). In the previous versions, OpenLiberty has implemented the specification in its own codebase. However, for MicroProfile Config 2.0 it consumes link:https://github.com/smallrye/smallrye-config[SmallRye's implementation] of the specification.
```

Author: Changed thanks for spotting.

MicroProfile Config 2.0 is a major version release (from 1.4) which introduces API/SPI changes and functional changes.

> ==== Functional Changes

Reviewer 2: The above line says Config 2.0 "introduces API/SPI changes and functional changes", however there are only headers for `Functional Changes` and `Incompatible changes `.

I would recommend either:
- Changing the above line to say "introduces many changes"
or
- Changing this title to say "API/SPI and functional changes"

Author: Changed to `API/SPI and functional changes`.

> * Provides a way to retrieve a number of related configuration properties with a specified prefix into a CDI bean, using the @link:https://github.com/eclipse/microprofile-config/blob/master/api/src/main/java/org/eclipse/microprofile/config/inject/ConfigProperties.java[ConfigProperties] annotation.

Reviewer: ```suggestion
* Provides a way to retrieve a number of related configuration properties with a specified prefix into a CDI bean, using the link:https://github.com/eclipse/microprofile-config/blob/master/api/src/main/java/org/eclipse/microprofile/config/inject/ConfigProperties.java[@ConfigProperties] annotation.
```

Ideally the at-sign should be part of the link since it is an annotation.

Author: Fixes, thanks.

[source, java]
----
@ConfigProperties(prefix="customer")
@Dependent
public static class BeanOne {
   private String name;
   int age;
}
----

> The code above retrieves the following defined properties for the `name` and `age` fields:

Reviewer 2: Is there any way to make the points that follow code blocks the same size and indentation as the bullet point they belong to? Currently, if a bullet point is split with a code block, the remaining text for the same bullet point does not look like it belongs to the same point.

Author: I've managed to do this on my local copy, will update after it has built on the draft site.

Author: Changed and is rendering correctly on draft.

Reviewer 2: Thanks that looks much better IMO. Good work!

The text after code blocks is still slightly bigger however. Is this something that could be fixed?

----
server.url=http://${server.host}/endpoint
server.host=example.org
----

*Note*: Previous working configurations may behave differently (if configuration in use contains values with Property Expressions syntax).

* Two new methods have been introduced to link:https://github.com/eclipse/microprofile-config/blob/master/api/src/main/java/org/eclipse/microprofile/config/Config.java[Config.java]:
** `Config.getValues(String propertyName, Class<T> propertyType)`
** `Config.getOptionalValues(String propertyName, Class<T> propertyType)`

The methods have been added to allow users to retrieve multi-valued properties as lists instead of arrays. The methods return the resolved property values with the specified `propertyType` for the specified `propertyName`.

* `OptinalInt`, `OptionalLong` and `OptionalDouble` are now provided as built-in link:https://download.eclipse.org/microprofile/microprofile-config-2.0-RC1/microprofile-config-spec.html#converter[Converters]. The new Converters can be used like any of the other built-in Converters, converting injected Config property values to a defined type:

[source, java]
----
@Inject
@ConfigProperty(name = "my.optional.int.property")

>   private OptionalInt optionalIntProperty;

Reviewer: ```suggestion
private OptionalInt optionalIntProperty;
```

removing extra spaces

Author: Fixed.

----

=== Incompatible changes

> If you move up from MicroProfile Config 1.x, 2.x, please take care of the following incompatible changes.

Reviewer: ```suggestion
If you move up from MicroProfile Config 1.x to 2.0, please take care of the following incompatible changes:
```

Author: Changed 👍

* `ConfigSource.getPropertyNames` is no longer a default method. The implementation of a `ConfigSource` must implement this method.

> * Previous versions of the specification would not evaluate Property Expressions. As such, previous working configuration may behave differently (if the used configuration contains values with Property Expressions syntax). Set the property `mp.config.property.expressions.enabled` with the value of `false`.

Reviewer 2: ```suggestion
* Previous versions of the specification would not evaluate Property Expressions. As such, previous working configuration may behave differently (if the used configuration contains values with Property Expressions syntax). Property Expressions can be disabled by setting the property `mp.config.property.expressions.enabled` with the value of `false`.
```

Author: Changed.

> * Empty values or other special characters are no longer valid Config property values. This means that `Config.getValue(String propertyName, Class<T> propertyType)` will never return `null`; a `NoSuchElementException` will be thrown if either a Config property value is not defined, is defined as an empty string, or its converter returns `null`. In the previous release, the empty value was returned as an empty value. For more information, please refer to this link:https://download.eclipse.org/microprofile/microprofile-config-2.0-RC1/microprofile-config-spec.html#_config_value_conversion_rules[section[].

Reviewer: ```suggestion
* Empty values or other special characters are no longer valid Config property values. This means that `Config.getValue(String propertyName, Class<T> propertyType)` will never return `null`; a `NoSuchElementException` will be thrown if either a Config property value is not defined, is defined as an empty string, or its converter returns `null`. In the previous release, the empty value was returned as an empty value. For more information, please refer to this link:https://download.eclipse.org/microprofile/microprofile-config-2.0-RC1/microprofile-config-spec.html#_config_value_conversion_rules[section].
```

Author: Fixed.

[source, java]
----
mp.jwt.verify.publickey.algorithm=ES256
----

The specification also adds support for the ES256 signature algorithm, while this Open Liberty beta supports using the RS384, RS512, HS384, HS512, ES256, ES384, and the ES512 signature algorithms.

To enable the the MicroProfile JWT 1.2 feature in your app, add the following to your `server.xml`:

[source, xml]
----
   <featureManager>
       <feature>mpJwt-1.2</feature>
   </featureManager>
----

For more information:

link:https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/release-notes.asciidoc#release-notes-for-microprofile-jwt-12[Release Notes for MicroProfile JWT 1.2]

link:https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/microprofile-jwt-auth-spec.asciidoc[MicroProfile JWT 1.2 Specification]

[#lra]
=== Support for MicroProfile LRA

> Long Running Actions (LRA) provide a cloud native transaction model based on a compensating Saga system. This allows loosely coupled services to coordinate long running activities, guaranteeing eventual data consistency without the need to put locks on data.

Reviewer: Do we have a link to the compensating Saga system?  It might be nice for people (like me) who are not familiar with it - or for people who want more details.

Author: @yongja79 I too have no idea what the compensating saga model is.

Reviewer 3: @austin0 Would [this](https://developer.ibm.com/depmodels/microservices/articles/use-saga-to-solve-distributed-transaction-management-problems-in-a-microservices-architecture/) be a suitable link? It describes the issue  then goes on to describe how both 2-phase commit and Saga's can solve the issue before saying way Saga's work better in a microservice situation

Author: I've added this and had it default to the `#saga` section.

The features provide a Coordinator, to which LRAs along with any corresponding method are registered, and the participant which allows annotated microservice methods to create a new or join an existing LRA.

> To start or join an LRA use the `@LRA` annotation. Depending on the type provided it may create a new LRA, join the existing LRA or continue outside one if none is provided. A full list of options can be found in the Microprofile Long Running Actions link:https://github.com/eclipse/microprofile-lra/blob/master/spec/src/main/asciidoc/microprofile-lra-spec.adoc[specifications].

Reviewer: ```suggestion
To start or join an LRA use the `@LRA` annotation. Depending on the type provided it may create a new LRA, join the existing LRA or continue outside one if none is provided. A full list of options can be found in the Microprofile Long Running Actions link:https://github.com/eclipse/microprofile-lra/blob/master/spec/src/main/asciidoc/microprofile-lra-spec.adoc[specification].
```

Reviewer: Also, this link points to the AsciiDoc version of the spec - it might be better to point to the generated HTML page at: https://download.eclipse.org/microprofile/microprofile-lra-1.0-M1/microprofile-lra-spec.html

Author: This has been fixed.

Author: Thanks for the link 👍

An example of always creating a new LRA when called would be:

[source, java]
----
    @LRA(LRA.Type.REQUIRES_NEW)
    @Path(“/start”)
    @PUT

>     public void startAction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)

Reviewer: ```suggestion
    public Response startAction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
```

This won't compile since it returns something for a void method. I suggest changing the return type to `Response` since that is what is returned in the method body.

Author: @yongja79 is this a correct change?

Author: Changed on Jason's OK 👍

[source, java]
----
    @Complete
    @Path("/complete")
    @PUT
    public Response completeAction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
    {
        /*
         * Business logic to cleanup resources once the LRA has closed successfully
         */
         return Response.ok().build();
    }  
----

Should a service in an LRA be canceled or timeout, the `@Compensate` methods for each participating service in the LRA will be called:

[source, java]
----
    @Compensate
    @Path("/compensate")
    @PUT
    public Response compensateAction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
    {
        /*
         * As the LRA has been canceled business logic to compensate any changes is necessary
         */
         return Response.ok().build();
    } 
----

To enable the LRA Coordinator feature in your app, add the following to your `server.xml`:
[source, xml]
----
<featureManager>
    <feature>cdi-2.0</feature>
    <feature>jaxrs-2.1</feature>
    <feature>mpLRACoordinator-1.0</feature>
</featureManager>
----

> To enable the LRA Participants feature in your app, add the following to your `server.xml`:

Reviewer: I'm not clear what the difference is between the coordinator and participants.  Maybe the LRA team could provide a short explanation so that users know which one they need depending on their use case?

Author: @yongja79

Reviewer 3: Added a suggested change to line 234 to cover this

Author: Suggestion added.

[source, xml]
----
<featureManager>
    <feature>cdi-2.0</feature>
    <feature>jaxrs-2.1</feature>
    <feature>mpLRA-1.0</feature>
</featureManager>
----

> For more information:

Reviewer: It might also be nice to mention that MP LRA is not part of the `microprofile-4.0` feature because it is a standalone specification project.  So unlike the other MP technologies mentioned above, users will need to explicitly add `mpLRA-1.0` or `mpLRACoordinator-1.0` to use it.

Author: Added a small note to make the reader aware of this.

[source,xml]
----
<dependency>
  <groupId>io.openliberty.beta</groupId>
  <artifactId>openliberty-runtime</artifactId>
  <version>20.0.0.12-beta</version>
  <type>pom</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty.beta', name: 'openliberty-runtime', version: '[20.0.0.12-beta,)'
}
----

Or take a look at our link:{url-prefix}/downloads/#runtime_betas[Downloads page].

[#jakarta]
== Jakarta EE 9 Beta Features package

The main change visible to developers in Jakarta EE is the names of packages changing to accommodate the new `jakarta.*` namespace. In this Open Liberty beta, we have a number of new API Release Candidates to join the expanding library of supported Jakarta packages.

This Open Liberty beta introduces the following Jakarta EE 9 features which now possess their all-new Jakarta EE 9 package names:

* Enterprise Beans Remote Client 2.0 (`ejbRemoteClient-2.0`)
* RESTful Web Services 3.0 (`restfulWS-3.0` and `restfulWSClient-3.0`)
* Jakarta Server Faces 3.0 (`faces-3.0`)

> * Connectors 2.0 (`connectors-2.0`)

Reviewer 4: Can be changed to:
```
Jakarta Connectors 2.0 (`connectors-2.0`)
```
To be consistent with the other feature's in this announcement.

Author: Added, thanks.

=== Try it now

To try out these Jakarta EE 9 features on Open Liberty in a lightweight package, just update your build tools to pull the Open Liberty Jakarta EE 9 Beta Features package instead of the main release. The beta works with Java SE 14, Java SE 11, or Java SE 8.

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty.beta</groupId>
    <artifactId>openliberty-jakartaee9</artifactId>
    <version>20.0.0.12-beta</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty.beta', name: 'openliberty-jakartaee9', version: '[20.0.0.12-beta,)'
}
----
Or take a look at our link:{url-prefix}/downloads/#runtime_betas[Downloads page].

[source, xml]
----
  <featureManager>
    <feature>jakartaee-9.0</feature>
  </featureManager>
----

Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 9 Web Profile beta features at once:

[source, xml]
----
  <featureManager>
    <feature>webProfile-9.0</feature>
  </featureManager>
----

> [#java15]

Reviewer 5: We are hoping to have official support for java 15 starting in 20.0.0.12.  If so, then this whole section would need to be moved from beta to the normal release blog.

Author: Yep, I don't see an issue with updating/duplicating this for the 20.0.0.12 GA.

== Support for Java SE 15

> Any official Java SE 15 release from link:https://adoptopenjdk.net[AdoptOpenJDK], OpenJDK, or Oracle works with Open Liberty. Java SE 15 is not a long-term supported release, with standard support scheduled to end in March 2021.

Reviewer: ```suggestion
Any official Java SE 15 release from link:https://adoptopenjdk.net/?variant=openjdk15&jvmVariant=openj9[AdoptOpenJDK], OpenJDK, or Oracle works with Open Liberty. Java SE 15 is not a long-term supported release, with standard support scheduled to end in March 2021.
```

Just a minor suggestion - the link can "pre-select" Java 15 from the AdoptOpenJDK site for the user.

Reviewer 5: I agree with Andy.  I also think we can link OpenJDK to the URL https://www.eclipse.org/openj9/ and Oracle to https://openjdk.java.net/groups/hotspot

Author: This change has been made 👍

[source, java]
----
mp.jwt.verify.publickey.algorithm=ES256
----

The specification also adds support for the ES256 signature algorithm, while this Open Liberty beta supports using the RS384, RS512, HS384, HS512, ES256, ES384, and the ES512 signature algorithms.

To enable the the MicroProfile JWT 1.2 feature in your app, add the following to your `server.xml`:

[source, xml]
----
   <featureManager>
       <feature>mpJwt-1.2</feature>
   </featureManager>
----

For more information:

link:https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/release-notes.asciidoc#release-notes-for-microprofile-jwt-12[Release Notes for MicroProfile JWT 1.2]

link:https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/microprofile-jwt-auth-spec.asciidoc[MicroProfile JWT 1.2 Specification]

[#lra]
=== Support for MicroProfile LRA

Long Running Actions (LRA) provide a cloud native transaction model based on a compensating Saga system. This allows loosely coupled services to coordinate long running activities, guaranteeing eventual data consistency without the need to put locks on data.

> The features provide a Coordinator, to which LRAs along with any corresponding method are registered, and the participant which allows annotated microservice methods to create a new or join an existing LRA.

Reviewer 3: ```suggestion
Two features have been provided for LRA support, a Coordinator feature and a Participant feature
The Coordinator is a dedicated service that participant microservices implicitly register their LRA’s through the use of annotations and has the responsibility of driving actions such as completing or compensating on an LRA. A Participatant is a JAX-RS microservice that wishes to participate in an LRA and has been annotated appropriately.

In order for an LRA to be used a single instance of the Coordinator feature must be enabled which will start a Coordinator application. Then a participating mircoservice must be written with the Participant feature enabled to allow the LRA annotations to be used.
```

    {
        /*
         * Business logic to cleanup resources once the LRA has closed successfully
         */
         return Response.ok().build();
    }  
----

Should a service in an LRA be canceled or timeout, the `@Compensate` methods for each participating service in the LRA will be called:

[source, java]
----
    @Compensate
    @Path("/compensate")
    @PUT
    public Response compensateAction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
    {
        /*
         * As the LRA has been canceled business logic to compensate any changes is necessary
         */
         return Response.ok().build();
    } 
----

To enable the LRA Coordinator feature in your app, add the following to your `server.xml`:
[source, xml]
----
<featureManager>
    <feature>cdi-2.0</feature>
    <feature>jaxrs-2.1</feature>
    <feature>mpLRACoordinator-1.0</feature>
</featureManager>
----

To enable the LRA Participants feature in your app, add the following to your `server.xml`:
[source, xml]
----
<featureManager>
    <feature>cdi-2.0</feature>
    <feature>jaxrs-2.1</feature>
    <feature>mpLRA-1.0</feature>
</featureManager>

> ----

Reviewer 3: ```suggestion

<lra port=“<port>” host=<hostname> path="lrac" />
----

where <port> is the default http port of the Open Liberty server that the Coordinator feature is enabled on and <hostname> is the host of that Open Liberty server
```

Author: I've gone ahead and made these changes.

---
layout: post
title: "MicroProfile Updates, New Jakarta EE 9 Packages and more in Open Liberty 20.0.0.12-beta"
categories: blog
author_picture: https://avatars3.githubusercontent.com/austin0
author_github: https://github.com/austin0
seo-title: MicroProfile Updates, New Jakarta EE 9 Packages and more in Open Liberty 20.0.0.12-beta - OpenLiberty.io
seo-description: OpenLiberty updates for a number of commonly used MicroProfile features and four new Jakarta EE 9 packages, now available in Open Liberty 20.0.0.12-beta.
blog_description: Updates for a number of commonly used MicroProfile features and four new Jakarta EE 9 packages.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---

> = MicroProfile Updates, New Jakarta EE 9 Packages and more in Open Liberty 20.0.0.12-beta

Reviewer 6: Change title to "More MicroProfile 4.0 previews, MicroProfile LRA, and more Jakarta EE 9 features in Open Liberty 20.0.0.12-beta".

Also change it in `title` and `seo-title` in the front matter/metadata.

Reviewer 6: Change `seo-description` and `description` to:

"Get previews of more MicroProfile 4.0 features (Config 2.0 and JWT 1.2), MicroProfile LRA, support for Java SE 15, and more Jakarta EE 9 features."

Austin Bailey <https://github.com/austin0>
:imagesdir: /
:url-prefix:
:url-about: /

> Open Liberty 20.0.0.12-beta comes complete with a range of previews for MicroProfile updates, including support for MicroProfile 4.0 and accompanying updates to MicroProfile Config (2.0) and MicroProfile JWT (1.2). Also available for preview are a multitude of Jakarta EE 9 features amongst which are changes to the Jakarta RESTful Web Services specification and the addition of message-driven bean support.

Reviewer 6: Nice first sentence but we can make it a bit more concise, and tweak the framing slightly to represent the MP 4.0 updates more accurately: "Open Liberty 20.0.0.12-beta comes with previews of a range of MicroProfile updates, and support for Java SE 15. The MicroProfile updates include more features in MicroProfile 4.0 (MicroProfile Config 2.0 and MicroProfile JWT 1.2), as well as the new MicroProfile LRA." Can you change it to that?

Author: Changed.

We have two beta packages for link:{url-about}[Open Liberty]:

* <<allbeta, All Beta Features>>: a larger package that contains all Open Liberty beta features (including Jakarta EE 9 beta features) and GA features and functions.
* <<jakarta, Jakarta EE 9 Beta Features>>: a lightweight package that contains only the Jakarta EE 9 features.

This means that you can now try out our in-development Open Liberty features by just adding the relevant coordinates to your build tools.

If you try either package, <<feedback, let us know what you think>>.
[#allbeta]
== All Beta Features package

The All Beta Features package includes the following beta features:

> * <<mp4, Support for MicroProfile 4.0>>

Reviewer 6: Change "Support for MicroProfile 4.0" to "New features  in MicroProfile 4.0".

Author: Changed.

> * <<conf, Support for MicroProfile Config 2.0>>

Reviewer 6: Nest Config and JWT under the previous MP 4.0 list item to show that they're part of MP 4.0.

Author: Changed.

* <<jwt, Support for MicroProfile JWT 1.2>>

> * <<lra, Support for MicroProfile LRA>>

Reviewer 6: (but leave the LRA bullet item as it is because it's not MP 4.0, even though it is MicroProfile too).

Author: Yep.

* <<java15, Support for Java SE 15>>

[#mp4]

> === Support for MicroProfile 4.0

Reviewer 6: Make this heading match the link in the list (see above for change).

Author: Done.

> The Open Liberty 20.0.0.12-beta provides most feature updates for MicroProfile 4.0-M1, this includes updates to MicroProfile Config, Fault Tolerance, Metrics, Health, Open Tracing, Open API and JWT, detailed below. While MicroProfile Rest Client is part of the MicroProfile 4.0-M1 specification, this feature narrowly missed the beta train and is scheduled for delivery in milestone 2.

Reviewer 6: Mostly fine but not entirely accurate. Can you change this para to:

"Open Liberty 20.0.0.12-beta provides another two feature updates for MicroProfile 4.0:

* MicroProfile Config 2.0
* MicroProfile JWT 1.2

Other MicroProfile 4.0 features were part of earlier betas: MicroProfile OpenAPI 2.0 in link:{url-prefix}/blog/2020/09/29/microprofile-openapi-beta-200011.html#openAPI[20.0.0.11-beta];  MicroProfile Fault Tolerance 3.0, MicroProfile Metrics 3.0, MicroProfile Health 3.0, and MicroProfile OpenTracing 2.0 in link:{url-prefix}/blog/2020/09/02/jakarta-microprofile-beta-200010.html#allbeta[20.0.0.10-beta].

While we'd hoped to also include MicroProfile Rest Client 2.0 (also part of MicroProfile 4.0), we didn't quite make it this time so it will make its debut in a future beta."

Make the Config and JWT list items into links to the relevant sections below. (We used to use lower level headings but we're already at an H3 here so it'll probably look odd to do that.)

Author: Done.

> MicroProfile 4.0 brings the following feature updates to Open Liberty:

Reviewer 6: Then remove this whole list.

Author: Removed.

* Config 2.0
* Fault Tolerance 3.0
* Metrics 3.0
* Health 3.0
* Open API 2.0
* JWT 1.2
* Open Tracing 2.0

> To enable the new beta features in your app, add them to your `pom.xml`:

Reviewer 6: Change this sentence to "To enable all the available MicroProfile 4.0 features at once, just add our MicroProfile 4.0 convenience feature to your `server.xml`:"

Author: Changed.

> [source, xml]

Reviewer 6: Replace this XML snippet with a normal server.xml snippet:

```
[source,xml]
----
   <featureManager>
       <feature>microProfile-4.0</feature>
   </featureManager>
----
```

Author: Changed.

----
    <dependencies>
        <dependency>
            <groupId>org.eclipse.microprofile</groupId>
            <artifactId>microprofile</artifactId>
            <version>4.0</version>
            <type>pom</type>
            <scope>provided</scope>
        </dependency>
    </dependencies>
----

For more information about the MicroProfile updates included in this milestone, visit the link:https://download.eclipse.org/microprofile/microprofile-4.0-M1/microprofile-spec-4.0-M1.html#microprofile4.0-M1[release page].

[#conf]

> === Support for MicroProfile Config 2.0

Reviewer 6: Change heading to "Move to Smallrye's MicroProfile Config implementation (MicroProfile Config 2.0)"

(use this format of heading for all updates: the key change then the feature name in parentheses, if applicable).

Author: Changed, and updated section directory to match.

> === Support for MicroProfile Config 2.0

Reviewer 6: Make the text in the link at the top of the page match this heading now.

> link:https://github.com/eclipse/microprofile-config[MicroProfile Config] is a specification which Open Liberty implements to allow application configuration from different locations (ConfigSources). In the previous versions, Open Liberty has implemented the specification in its own codebase. However, for MicroProfile Config 2.0 it consumes link:https://github.com/smallrye/smallrye-config[SmallRye's implementation] of the specification.

Reviewer 6: Replace "is a specification which Open Liberty implements to allow" with "enables"
(almost all of Liberty is an implementation of some specification or another - that's not important or useful to the reader)

Reviewer 6: " In the previous versions, Open Liberty has implemented the specification in its own codebase. However, for MicroProfile Config 2.0 it consumes link:https://github.com/smallrye/smallrye-config[SmallRye's implementation] of the specification."

change to

"MicroProfile Config 2.0 in Open Liberty consumes link:https://github.com/smallrye/smallrye-config[SmallRye's implementation of MicroProfile Config] instead of implementing our own codebase."

Author: Changed.

> MicroProfile Config 2.0 is a major version release (from 1.4) which introduces API/SPI and functional changes.

Reviewer 6: "introduces functional changes and API/SPI incompatible changes." (match the order of the section that follow)

> ==== Functional Changes

Reviewer 6: "Functional changes" (lower-case c).

In fact, go with "Functional changes in Microprofile Config 2.0" - better for skim-reading and for SEO.

Author: Changed.

+
[source, java]
----
@ConfigProperties(prefix="customer")
@Dependent
public static class BeanOne {
   private String name;
   int age;
}
----
+
The code above retrieves the following defined properties for the `name` and `age` fields:
+
[source, java]
----
customer.name=Bob
customer.age=24
----

* Allows you to easily determine which was the "winning" `ConfigSource` (the `ConfigSource` with the highest ordinal) for a specified property via link:https://github.com/eclipse/microprofile-config/blob/master/api/src/main/java/org/eclipse/microprofile/config/ConfigValue.java[ConfigValue]. A `ConfigValue` object holds information about the configuration property name, configuration value, the ConfigSource name from where the configuration property was loaded and the ordinal of the `ConfigSource`. You can access these values with the following method:
+
[source, java]
----
ConfigValue cv = Config.getConfigValue(String propertyName);
cv.getSourceName() // the config source that supplies the configure value for the above property
----

> * Enables an application to include configuration for different environments and development stages while only one of them is active (e.g. `dev`, `test`, `prod`) via Configuration Profiles. The Config profile can be specified via the `mp.config.profile` property, which can be set in any of the configuration sources. Once it is set, the corresponding properties associated with the active profile are used. The value can also be set at application start-up, for example:

Reviewer 6: Change "via Configuration Profiles" to "using configuration profiles"

> * Enables an application to include configuration for different environments and development stages while only one of them is active (e.g. `dev`, `test`, `prod`) via Configuration Profiles. The Config profile can be specified via the `mp.config.profile` property, which can be set in any of the configuration sources. Once it is set, the corresponding properties associated with the active profile are used. The value can also be set at application start-up, for example:

Reviewer 6: "via" to "using"

Reviewer 6: "The Config profile" to "The config profile"

+
[source, xml]
----
java -jar myapp.jar -Dmp.config.profile=testing
----
+
The `mp.config.profile` property works on both the property level and config file level:
+

> On the property level profiles can be utilised by setting Config property values specific to the selected profile. For example, the `vehicle.name` property in the following example would be set to `bike` from the previous Config start up command:

Reviewer 6: "utilised" to "used"

Reviewer 6: "start up command" to "start-up command"

+
[source, java]
----
%dev.vehicle.name=car
%testing.vehicle.name=bike
vehicle.name=lorry
----
+
If `mp.config.profile` was not set, `vehicle.name` would be set to `lorry`.
+
On the config file (`microprofile-config.property`) level the property `mp.config.profile` also activates the corresponding `microprofile-config-<mp.config.profile>.properties` with `microprofile-config.properties` as the fallback.

* Supports link:https://download.eclipse.org/microprofile/microprofile-config-2.0-RC1/microprofile-config-spec.html#property-expressions[Property Expressions], which provide a way to set and expand variables in property values. For example, in the following configuration properties, the property `server.url` will be resolved to `http://example.org/endpoint`:
+
[source, java]
----
server.url=http://${server.host}/endpoint
server.host=example.org
----
+

> *Note*: Previous working configurations may behave differently (if configuration in use contains values with Property Expressions syntax).

Reviewer 6: Change to "Previous working configurations might behave differently if the configuration in use contains values with Property Expressions syntax."

* Two new methods have been introduced to link:https://github.com/eclipse/microprofile-config/blob/master/api/src/main/java/org/eclipse/microprofile/config/Config.java[Config.java]:
** `Config.getValues(String propertyName, Class<T> propertyType)`
** `Config.getOptionalValues(String propertyName, Class<T> propertyType)`
+

> The methods have been added to allow users to retrieve multi-valued properties as lists instead of arrays. The methods return the resolved property values with the specified `propertyType` for the specified `propertyName`.

Reviewer 6: "to allow users" to "to enable you"

* `OptinalInt`, `OptionalLong` and `OptionalDouble` are now provided as built-in link:https://download.eclipse.org/microprofile/microprofile-config-2.0-RC1/microprofile-config-spec.html#converter[Converters]. The new Converters can be used like any of the other built-in Converters, converting injected Config property values to a defined type:
+
[source, java]
----
@Inject
@ConfigProperty(name = "my.optional.int.property")
private OptionalInt optionalIntProperty;
----

=== Incompatible changes

If you move up from MicroProfile Config 1.x to 2.0, please take care of the following incompatible changes:

* `ConfigSource.getPropertyNames` is no longer a default method. The implementation of a `ConfigSource` must implement this method.

> * Previous versions of the specification would not evaluate Property Expressions. As such, previous working configuration may behave differently (if the used configuration contains values with Property Expressions syntax). Property Expressions can be disabled by setting the property `mp.config.property.expressions.enabled` with the value of `false`.

Reviewer 6: Change "Property Expressions" to "property expressions" throughout this para.

> * Empty values or other special characters are no longer valid Config property values. This means that `Config.getValue(String propertyName, Class<T> propertyType)` will never return `null`; a `NoSuchElementException` will be thrown if either a Config property value is not defined, is defined as an empty string, or its converter returns `null`. In the previous release, the empty value was returned as an empty value. For more information, please refer to this link:https://download.eclipse.org/microprofile/microprofile-config-2.0-RC1/microprofile-config-spec.html#_config_value_conversion_rules[section].

Reviewer 6: "[section]" to "[section of the spec]"

To enable the new beta features in your app, add them to your `server.xml`:

[source, xml]
----
<featureManager>
   <feature>mpConfig-2.0</feature>
</featureManager>
----

For more information:

> link:https://download.eclipse.org/microprofile/microprofile-config-2.0-RC1/microprofile-config-spec.html#release_notes_20[Release Notes for MicroProfile Config 2.0]

Reviewer 6: Make these two links into a bulleted list so it's more compact.

link:https://github.com/eclipse/microprofile-config/milestone/8?closed=1[MicroProfile Config 2.0 Milestone]

[#jwt]

> === Support for MicroProfile JWT 1.2

Reviewer 6: Change heading to "New configuration properties for MicroProfile JWT (MicroProfile JWT 1.2)"

And make sure the link at the top of the page matches.

> The MicroProfile JSON Web Token 1.2 specification allows using a JWT token for authenticating and authorizing requests to a service. The specification simplifies the configuration for managing the validation of the JWT by introducing new MicroProfile Config properties. Enhanced signature algorithm support is added in this Open Liberty beta.

Reviewer 6: Change to "With MicroProfile JSON Web Token (JWT) 1.2 you can use a JWT token for authenticating and authorizing requests to a service. MicroProfile JWT 1.2 simplifies the configuration for managing the validation of the JWT by introducing new MicroProfile Config properties. Enhanced signature algorithm support is added in this Open Liberty beta."

> Version 1.2 of the MicroProfile JSON Web Token specification adds the following MicroProfile Config properties to control different aspects of the JWT validation.

Reviewer 6: Change to "MicroProfile JSON Web Token 1.2 adds the following MicroProfile Config properties to control different aspects of the JWT validation."

(basically we rarely need to talk about "specification". It's just what the feature does. The fact that it's been implemented from a specification is an internal detail for us to care about, not for developers who use it to care about.)

==== mp.jwt.token.header

The `mp.jwt.token.header` property allows you to control the HTTP request header which is expected to contain a JWT token. You can specify either Authorization (default) or Cookie values.

==== mp.jwt.token.cookie

The `mp.jwt.token.cookie` property allows you to specify the name of the cookie which is expected to contain a JWT token. The default value is `Bearer` if not specified.

==== mp.jwt.verify.audiences

> The `mp.jwt.verify.audiences` property allows you to create list of allowable audience (`aud`) values. At least one of these must be found in the claim. Previously, this had to be configured in the `server.xml` file. Now, you can configure the audiences in the MicroProfile Config property as follows:

Reviewer 6: "a list"

> [source, java]

Reviewer 6: think this should just be "text" not "java" as these are entries in the config properties file.

----
mp.jwt.verify.audiences=conferenceService
----

==== mp.jwt.verify.publickey.algorithm

The `mp.jwt.verify.publickey.algorithm` property allows you to control the Public Key Signature Algorithm that is supported by the MicroProfile JWT endpoint. The default value is RSA256 if not specified. Previously, this had to be configured in the `server.xml` file. Now, you can configure the public key algorithm used for verification of the JWT in the MicroProfile Config property as follows:

> [source, java]

Reviewer 6: text not java in the properties file

----
mp.jwt.verify.publickey.algorithm=ES256
----

The specification also adds support for the ES256 signature algorithm, while this Open Liberty beta supports using the RS384, RS512, HS384, HS512, ES256, ES384, and the ES512 signature algorithms.

> To enable the the MicroProfile JWT 1.2 feature in your app, add the following to your `server.xml`:

Reviewer 6: "add to following" to "add the feature to"

[source, xml]
----
   <featureManager>
       <feature>mpJwt-1.2</feature>
   </featureManager>
----

For more information:

> link:https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/release-notes.asciidoc#release-notes-for-microprofile-jwt-12[Release Notes for MicroProfile JWT 1.2]

Reviewer 6: Make the links into a bulleted list.

link:https://github.com/eclipse/microprofile-jwt-auth/blob/master/spec/src/main/asciidoc/microprofile-jwt-auth-spec.asciidoc[MicroProfile JWT 1.2 Specification]

[#lra]

> === Support for MicroProfile LRA

Reviewer 6: Change heading to "Enable loosely-coupled services to coordinate long running activities (MicroProfile LRA Participants 1.0 & MicroProfile LRA Coordinator 1.0)"

And make sure the link list at the top of the post matches.

> Long Running Actions (LRA) provide a cloud native transaction model based on a link:https://developer.ibm.com/depmodels/microservices/articles/use-saga-to-solve-distributed-transaction-management-problems-in-a-microservices-architecture#saga[compensating Saga system]. This allows loosely coupled services to coordinate long running activities, guaranteeing eventual data consistency without the need to put locks on data.

Reviewer 6: "cloud-native transaction model"

Two features have been provided for LRA support, a Coordinator feature and a Participant feature:

> * The Coordinator is a dedicated service that participant microservices implicitly register their LRA’s through the use of annotations and has the responsibility of driving actions such as completing or compensating on an LRA. 

Reviewer 6: "that participant" to "with which participant"

"...annotations and has..." to "...annotations. It has..."

> * A Participatant is a JAX-RS microservice that wishes to participate in an LRA and has been annotated appropriately.

Reviewer 6: "Participant" (typo)

    {
        /*
         * Business logic that may call upon other services to participate in this Long Running Action
         */
         return Response.ok().build();
    } 
----

Once the LRA has been successfully finished all participating services will have their `@Complete` annotated methods called to perform the necessary cleanup:

[source, java]
----
    @Complete
    @Path("/complete")
    @PUT
    public Response completeAction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
    {
        /*
         * Business logic to cleanup resources once the LRA has closed successfully
         */
         return Response.ok().build();
    }  
----

Should a service in an LRA be canceled or timeout, the `@Compensate` methods for each participating service in the LRA will be called:

[source, java]
----
    @Compensate
    @Path("/compensate")
    @PUT
    public Response compensateAction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
    {
        /*
         * As the LRA has been canceled business logic to compensate any changes is necessary
         */
         return Response.ok().build();
    } 
----

> To enable the LRA Coordinator feature in your app, add the following to your `server.xml`:

Reviewer 6: "add the following" to "add the feature"

[source, xml]
----
<featureManager>
    <feature>cdi-2.0</feature>
    <feature>jaxrs-2.1</feature>
    <feature>mpLRACoordinator-1.0</feature>
</featureManager>
----

> To enable the LRA Participants feature in your app, add the following to your `server.xml`:

Reviewer 6: "add the following" to "add the feature"

[source, xml]
----
<featureManager>
    <feature>cdi-2.0</feature>
    <feature>jaxrs-2.1</feature>
    <feature>mpLRA-1.0</feature>
</featureManager>

<lra port=“<port>” host=<hostname> path="lrac" />
----

Where `<port>` is the default `HTTP` port of the Open Liberty server with the Coordinator feature is enabled, and `<hostname>` is the hostname provided for that Open Liberty server.

> It should be noted that MicroProfile Long Running Actions is not a part of the <<mp4, MicroProfile 4.0>> specification, and so needs to be explicitly added as a feature to be used.

Reviewer 6: Remove "It should be noted that" (not necessary)

For more information:

> link:https://github.com/eclipse/microprofile-lra/blob/master/spec/src/main/asciidoc/microprofile-lra-spec.adoc[Microprofile Long Running Actions Specification]

Reviewer 6: Make this a single item bulleted list - just for consistency with the others.

* Jakarta Server Pages 3.0 (`jsp-3.0`)
* Jakarta Expression Language 4.0 (`el-4.0`)

Enable the Jakarta EE 9 beta features in your app's `server.xml`. You can enable the individual features you want or you can just add the Jakarta EE 9 convenience feature to enable all of the Jakarta EE 9 beta features at once:

=== Try it now

To try out these Jakarta EE 9 features on Open Liberty in a lightweight package, just update your build tools to pull the Open Liberty Jakarta EE 9 Beta Features package instead of the main release. The beta works with Java SE 15, Java SE 11, or Java SE 8.

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty.beta</groupId>
    <artifactId>openliberty-jakartaee9</artifactId>
    <version>20.0.0.12-beta</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty.beta', name: 'openliberty-jakartaee9', version: '[20.0.0.12-beta,)'
}
----
Or take a look at our link:{url-prefix}/downloads/#runtime_betas[Downloads page].

> [source, xml]

Reviewer 6: You've grabbed the dodgy version of Jakub's template (he's fixing it).
Move lines 412 (`[source,xml]`) to 426 (`----`) up to immediately before the `=== Try it now` heading (just after the paragraph that starts "Enable the Jakarta EE 9 features...".

This section is the server.xml entries that you use to enable the Jakarta EE 9 features. They shouldn't be down here after the Maven etc info.

----
  <featureManager>
    <feature>jakartaee-9.0</feature>
  </featureManager>
----

Or you can add the Web Profile convenience feature to enable all of the Jakarta EE 9 Web Profile beta features at once:

[source, xml]
----
  <featureManager>
    <feature>webProfile-9.0</feature>
  </featureManager>
----

[#java15]

> == Support for Java SE 15

Reviewer 6: Can you move this whole Java 15 section to just after the LRA section (as indicated in the link list at the top of that section)? It's part of the All Beta Features package. (I know it's technically part of the JEE 9 one too but that already mentions it and doesn't need to give any more info about it.)

Keep in mind, Eclipse OpenJ9 link:{url-prefix}/blog/2019/10/30/faster-startup-open-liberty.html[typically offers faster startup times] than Hotspot.

The primary features added in this release include:

* link:https://openjdk.java.net/jeps/379[JEP 379] Shenandoah: A Low-Pause-Time Garbage Collector
* link:https://openjdk.java.net/jeps/377[JEP 377] ZGC: A Scalable Low-Latency Garbage Collector
* link:https://openjdk.java.net/jeps/378[JEP 378] Text Blocks
* link:https://openjdk.java.net/jeps/384[JEP 384] Records (Second Preview)
* link:https://openjdk.java.net/jeps/360[JEP 360] Sealed Classes (Preview)

For more information on downloading a version of Java 15, please visit link:https://adoptopenjdk.net/index.html?variant=openjdk15&jvmVariant=openj9[AdoptOpenJDK.net], link:https://www.eclipse.org/openj9/[Eclipse.org] or link:https://openjdk.java.net/groups/hotspot[OpenJDK.java.net].

> For working with the `server.env` file in Open Liberty, please visit the `Configuration Files` section of the Open Liberty link:{url-prefix}/docs/latest/reference/config/server-configuration-overview.html[Server Configuration Overview documentation].

Reviewer 6: "please visit the" to "see the" (we don't need to say please in this kind of info)

> For more information on new features available in Java 15, please visit link:https://openjdk.java.net/projects/jdk/15/[OpenJDK].

Reviewer 6: "please visit" to "see" (again, no pleases)

[#java15]
== Support for Java SE 15

> Any official Java SE 15 release from link:https://adoptopenjdk.net/?variant=openjdk15&jvmVariant=openj9[AdoptOpenJDK], OpenJDK, or Oracle works with Open Liberty. Java SE 15 is not a long-term supported release, with standard support scheduled to end in March 2021.

Reviewer 5: I agree with Andy. I also think we can link OpenJDK to the URL https://www.eclipse.org/openj9/ and Oracle to https://openjdk.java.net/groups/hotspot

Author: Could definitely reword it as OpenJDK doesn't produce any binaries one their own (afaik). Oracle could be linked to https://jdk.java.net/15/ for sure though.

Author: Added the preset link for JDK15/J9 and added a link to Oracle JDKs.
